src/main/java/com/xebialabs/overthere/smb/SmbConnection.java

```diff
@@ -0,0 +1,211 @@
+ * of the License.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ * <p>
+ * You should have received a copy of the GNU General Public License along with this
+ * program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth
+ * Floor, Boston, MA 02110-1301  USA
+ */
+package com.xebialabs.overthere.smb;
+
+import com.hierynomus.ntlm.NtlmException;
+import com.hierynomus.smbj.DefaultConfig;
+import com.hierynomus.smbj.SMBClient;
+import com.hierynomus.smbj.auth.AuthenticationContext;
+import com.hierynomus.smbj.connection.Connection;
+import com.hierynomus.smbj.session.Session;
+import com.hierynomus.smbj.share.DiskShare;
+import com.xebialabs.overthere.ConnectionOptions;
+import com.xebialabs.overthere.OverthereFile;
+import com.xebialabs.overthere.RuntimeIOException;
+import com.xebialabs.overthere.cifs.CifsConnectionType;
+import com.xebialabs.overthere.proxy.ProxyConnection;
+import com.xebialabs.overthere.spi.AddressPortMapper;
+import com.xebialabs.overthere.spi.BaseOverthereConnection;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.security.Security;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.xebialabs.overthere.ConnectionOptions.*;
+import static com.xebialabs.overthere.cifs.CifsConnectionBuilder.CONNECTION_TYPE;
+import static com.xebialabs.overthere.smb.SmbConnectionBuilder.*;
+import static java.net.InetSocketAddress.createUnresolved;
+
+public class SmbConnection extends BaseOverthereConnection {
+
+    private final SMBClient client;
+    private final String hostname;
+    private final int smbPort;
+    private Connection connection;
+    private Session session;
+    private int port;
+    private Map<String, DiskShare> shareCache = new ConcurrentHashMap<>();
+
+    protected final String password;
+    protected CifsConnectionType cifsConnectionType;
+    protected final String username;
+
+    static {
+        if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME) == null) {
+            Security.addProvider(new BouncyCastleProvider());
+        }
+    }
+
+    protected SmbConnection(String protocol, ConnectionOptions options, AddressPortMapper mapper, boolean canStartProcess) {
+        super(protocol, options, mapper, canStartProcess);
+        this.cifsConnectionType = options.getEnum(CONNECTION_TYPE, CifsConnectionType.class);
+        if (mapper instanceof ProxyConnection) {
+            throw new IllegalArgumentException("Cannot open a smb:" + cifsConnectionType.toString().toLowerCase() + ": connection through an HTTP proxy");
```

Comment:
I have not tested it myself, we have a similar existing check for CIFS connection, I believe same is applicable to SMB as well